### ISSUE TITLE
Create relations between accumulate_grad and operations in post_accumulate_grad_hook

### ISSUE THREAD

**Issue by fegin** [pytorch/pytorch#119906]
### 🚀 The feature, motivation and pitch

When using `compiled_autograd` with `register_post_accumulate_grad_hook()`, there are no relations between the nodes of `accumuldate_grad` and the ops in `post_accumulate_grad_hook` in the traced FX graph. If we use this traced FX graph to do bucketing and communication reordering, there are chances that what happen in `post_accumulate_grad_hook` will be reordered to before `post_accumulate_grad_hook`

In this example:
```
import torch
from torch import _inductor as inductor
from torch._dynamo import compiled_autograd

def compiler_fn(gm):
    def inner(gm_, example_inputs_):
        print(gm_)
        return inductor.compile(gm_, example_inputs_)
    return torch.compile(gm, fullgraph=True, backend=inner)

def hook(param):
    # Mimic what DDP does.
    #
    # gradient = param.grad / self.process_group.size()
    # gradient = fcol.all_reduce(gradient, "sum", self.process_group)
    # param.grad.copy_(gradient)

    all_reduce_result = param.grad / 2
    param.grad.copy_(all_reduce_result)

x = torch.ones(10)
x.requires_grad = True
def run(input):
    return x * input

x.register_post_accumulate_grad_hook(hook)
with compiled_autograd.enable(compiler_fn):
    for i in range(5):
        run(i).sum().backward()
        # Mimic optimizer.zero_grad() to clear the gradient
        print(x.grad)
        x.grad = None
```

The `inner` printed out the fx graph as the following graph:
```
def forward(self, L_inputs_0_ : torch.Tensor, L_inputs_1_ : torch.Tensor, L_inputs_2_ : torch.Tensor, L_inputs_2_grad : torch.Tensor):
    getitem = L_inputs_0_
    getitem_1 = L_inputs_1_
    call_hook = L_inputs_2_
    l_inputs_2_grad = L_inputs_2_grad
    expand = torch.ops.aten.expand.default(getitem, [10]);  getitem = None
    mul = torch.ops.aten.mul.Tensor(expand, getitem_1);  expand = getitem_1 = None
    accumulate_grad__default = torch.ops.inductor.accumulate_grad_.default(call_hook, mul);  call_hook =
mul = None
    all_reduce_result = l_inputs_2_grad / 2
    copy_ = l_inputs_2_grad.copy_(all_reduce_result);  l_inputs_2_grad = all_reduce_result = None
    return ()
```

The input arguments of `accumulate_grad_` is from `L_inputs_2_` while the the argument of `all_reduce_result` is from `L_inputs_2_grad`. And this is risky because the optimization can reorder `all_reduce_result` earlier than `accumulate_grad_` because its only dependency is `L_inputs_2_grad`.

Is there any way we can trace the relation of the two ops?

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519

**Comment by bdhirsh** [comment]
One thought: if `inputs_2.grad` is `None` before the backward executes, it seems a bit wrong for the produced graph to take in `inputs_2.grad` as a graph argument.

@anijain2305 [pointed out](https://github.com/pytorch/pytorch/issues/118435#issuecomment-1943241652) in the other issue the reason we see guards on `inpts_2.grad`. We also end up with `inpts_2_grad` as an extra GraphArg because of this code: https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/builtin.py#L1301.

One problem is that a parameter might have a `None` grad before the graph executes, but its grad will be filled part-way through execution of the graph, due to the `AccumulateGrad` node.

It might help the discussion if we're in agreement on what the graph should look like.

Would a graph like this be better?
```
def forward(self, inp0, inp1, inp2):
    expand = torch.ops.aten.expand.default(inp0, [10])
    mul = torch.ops.aten.mul.Tensor(expand, inp1)
    inp2_grad = torch.ops.inductor.accumulate_grad_.default(inp2, mul)
    all_reduce_result = inp2_grad / 2
    copy_ = inp2_grad.copy_(all_reduce_result)
    return ()
```

This graph seems nicer because it solves the ordering issue mentioned above: `all_reduce_result` is now forced to run after the  `accumulate_grad` node, because its input is an output of accumulate_grad.

A few consequences of this graph:

(1) It would require the existing `accumulate_grad_(x)` operator to return `x.grad` as an output (making this op pseudo-functional, except for the fact that it also secretly mutates `x.grad`)

(2) When dynamo witnesses this code:
```
inp2_grad = torch.ops.inductor.accumulate_grad_.default(inp2, mul)
```

It will make a `TensorVariable` for `inp2_grad`. Dynamo would also (somehow) need to know that on any future accesses of `inp2.grad`, there is an existing `TensorVariable` for `inp2.grad` that it should return instead (the proxy returned by `accumulate_grad_`). Maybe we can do this by having dynamo do two things:

(a) special-case `accumulate_grad_(x, y)` (to mark `x` as having a `.grad` field)
(b) special-case `getattr(x, "grad")` (to see if "x" was already an input to an accumulate_grad_ call, so we have an existing TensorVariable handy for its grad, and don't need to generate one).

(3) It technically doesn't need to specialize on whether `inp.grad` was None or not before execution: since these details are hidden in the implementation of `accumulate_grad_`

**Comment by ezyang** [comment]
fyi there's also parallel discussion happening at https://github.com/pytorch/pytorch/issues/118435

**Comment by fegin** [comment]
@bdhirsh The graph you proposed looks nice. The graph explains the correct dependency. The `all_reduce` actually consumes the result from `accumulate_grad_` even though `accumuldate_grad_` is not a functional ops now.

**Comment by masnesral** [comment]
@bdhirsh I'm marking this as triaged since the parallel discussion Ed links is triaged, but please update the labels if this should be discussed at the triage review meeting.
